Initial import of "Protect an API in Azure Functions - Python - With EasyAuth" (#10)
Azure-Samples/ms-identity-docs-code-python: src/protect-function-app/README.md

products:
- azure
- azure-active-directory
- azure-functions
urlFragment: ms-identity-docs-code-functions-python
---
-->
# Python | Azure Function | protect an API | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This Python Azure Function protects its own API (HTTP trigger) with the combination of the built-in authentication & authorization feature of Azure Functions (commonly known as "Easy Auth") and JWT access token scope validation.

```console
$ curl https://<your-function>.azurewebsites.net/api/greeting -H "Authorization: Bearer <valid-access-token>"
Hello, world. You were able to access this because you provided a valid access token with the Greeting.Read scope as a claim.
```

> :page_with_curl: This sample application backs one or more technical articles on docs.microsoft.com. <!-- TODO: Link to first tutorial in series when published. -->

## Prerequisites

> - Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.

Reviewer: nit: I think it is ok to mention that now users are going to need a subscription as well.

Author: Wouldn't saying that they need a deployed Azure Function cover that requirement already?  What value would adding a _pre-_ prereq bring to this list?

Reviewer: The idea behind this is that I think it is common patter to introduce our users with something as shown below:

```bash
1. An Azure subscription. You can [open an account for free](https://azure.microsoft.com/free).
```

But yeah, asking for an Azure Function resource instance already deployed I think it is something that will enable our users to immediately infer this requirement.

Reviewer: Another option could be to cross link to some another doc like the one for register apps while mentioning the need for an already-deployed Function App. 

This option looks more appealing to the README style we are following I guess.

Author: Just chatted with @ferantivero -- all good to close out this comment.  The already existing pre-req covers all suggestions above.

- A missing or invalid (expired, wrong audience, etc) token will result in a `401` response. (Handled by Azure Functions authentication)
- An otherwise valid token without the proper scope will result in a 403 response.
- A valid token with the proper scope of `Greeting.Read` will result in the "Hello, world" message.

### Running locally

At the time of this writing, Function app authentication does not support a local development experience that has parity with the on-Azure runtime. You can still execute this locally with `func start` but the authentication functionality provided by the Function app service on Azure will not be invoked; all JWT token validation for authorization (signature, iss, exp, aud) will be skipped.

#### Web server gateway interface (WSGI) alternative

> In Python, it is common to use a web framework, such as Django or Flask, to handle APIs. While this sample doesn't show it, Azure Functions for Python does support [WSGI integration](https://docs.microsoft.com/python/api/azure-functions/azure.functions.wsgimiddleware), allowing you to use the web framework as your HTTP request pipeline. Introducing WSGI into this scenario is out of scope for this introduction, but doing so would allow you to use framework-specific authentication mechanisms that might feel more native for your Python API.

Reviewer 2: Love this - nice work! 🥇

Author: 😊

| **Supported account types**    | **Accounts in this organizational directory only (Single tenant)**   | Suggested value for this sample.                                                |
| **Platform type**              | _None_                                                               | No redirect URI required; don't select a platform.                              |
| **Scopes defined by this API** | **Scope name**: `Greeting.Read`<br/>**Who can consent?**: **Admins and users**<br/>**Admin consent display name**: `Read API Greetings`<br/>**Admin consent description**: `Allows the user to see greetings from the API.`<br/>**User consent display name**: `Read API Greetings`<br/>**User consent description**: `Allows you to see greetings from the API.`<br/>**State**: **Enabled** | Required scope for this sample. |

> :information_source: **Bold text** in the table matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

> ### 2. Enable Function App Authentication
> 
> 1. Open the existing Function App in the Azure portal.
> 1. Click on **Authentication** and then **Add identity provider**.
> 1. Use these settings in your identity provider configuration.

Reviewer 2: ```suggestion
### 2. Enable Function app authentication

Next, complete the steps in [Enable Azure Active Directory in your App Service app](https://docs.microsoft.com/azure/app-service/configure-authentication-provider-aad?toc=/azure/azure-functions/toc.json#-enable-azure-active-directory-in-your-app-service-app) to add Azure Active Directory as an identity provider for your API.
```

Let's let the Functions/App Service docs do the heavy lifting on any procedural stuff for those services. Whenever possible, avoid steps in the README sections.

Author: Oh, yeah, great idea. I'll tidy this up.

   | Identity provider setting       | Value for this sample app                                | Notes                                                                            |
   |--------------------------------:|:---------------------------------------------------------|:---------------------------------------------------------------------------------|
   | **Identity provider**           | **Microsoft**                                            | Required value for this sample.                                                  |
   | **App registration type**       | **Provide the details of an existing app registration**  | Required value for this sample.                                                  |

>    | **Application (client) ID**     | `<client-id>`                                            | Required value for this sample. <br/> 'Application (client) ID' of app registration in Azure portal - this value is a GUID               |

Reviewer 2: ```suggestion
   | **Application (client) ID**     | `<client-id>`                                            | Required value for this sample. <br/> 'Application (client) ID' of the API's app registration in Azure portal - this value is a GUID               |
```

Author: Added verbatium.

   | **Client secret (recommended)** | _None_                                                   | Suggested value for this sample. <br/> This sample doesn't require this feature. |

>    | **Issuer URL**                  | `https://sts.windows.net/<Tenant ID>/v2.0`               | Required value for this sample. <br/> Update to include 'Tenant ID' of your Azure AD instance - this value is a GUID                     |

Reviewer 2: This looks suspect, I'd expect it to be something like:

`https://login.microsoftonline.com/<tenant-id>/v2.0`

Which jives with the App Service auth doc:

![image](https://user-images.githubusercontent.com/13394541/151084154-18312665-150d-4b44-95c7-41067dbccd87.png)

Author: Yeah, they are both valid.  The sts just happens to be what the portal will fill in automatically in most situations on this UI.  But you're right that we should lead with the login.microsoft.com DNS value instead.  I should have picked up on that in the first pass.  Thanks, will update

   | **Allowed token audiences**     | `api://<client-id>`                                      | Required value for this sample. <br/> 'Application ID URI' of app registration in Azure portal - this value typically starts with api:// |
   | **Restrict access**             | **Require authentication**                               | Required value for this sample.                                                  |
   | **Unauthenticated requests**    | **HTTP 401 Unauthorized: recommended for APIs**          | Suggested value for this sample.                                                 |
   | **Token store**                 | _Unselected_                                             | Suggested value for this sample. |

   > :information_source: **Bold text** in the table matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

> 
> 1. Click **Add**.

Reviewer 2: ```suggestion
```

Author: Removed in Simplification applied in https://github.com/Azure-Samples/ms-identity-docs-code-python/pull/10/commits/852d23cbe8069ceaf40758dec86e08a12e509063

> ### 3. Deploy the Function App

Reviewer 2: ```suggestion
### 3. Deploy the Function app
```

Author: Updated here and everywhere.  The portal shows it as "Function App" -- so I went with that.  I should have checked the docs product naming list to cross check!

Updated in https://github.com/Azure-Samples/ms-identity-docs-code-python/pull/10/commits/852d23cbe8069ceaf40758dec86e08a12e509063
